app/scripts/migrations/127.ts

```diff
@@ -0,0 +1,265 @@
+  'eth_gasPrice',
+  'eth_getBalance',
+  'eth_getBlockByHash',
+  'eth_getBlockByNumber',
+  'eth_getBlockReceipts',
+  'eth_getBlockTransactionCountByHash',
+  'eth_getBlockTransactionCountByNumber',
+  'eth_getCode',
+  'eth_getFilterChanges',
+  'eth_getFilterLogs',
+  'eth_getLogs',
+  'eth_getProof',
+  'eth_getStorageAt',
+  'eth_getTransactionByBlockHashAndIndex',
+  'eth_getTransactionByBlockNumberAndIndex',
+  'eth_getTransactionByHash',
+  'eth_getTransactionCount',
+  'eth_getTransactionReceipt',
+  'eth_getUncleCountByBlockHash',
+  'eth_getUncleCountByBlockNumber',
+  'eth_maxPriorityFeePerGas',
+  'eth_newBlockFilter',
+  'eth_newFilter',
+  'eth_newPendingTransactionFilter',
+  'eth_sendRawTransaction',
+  'eth_syncing',
+  'eth_uninstallFilter',
+];
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 127;
+
+/**
+ * This migration transforms `eth_accounts` and `permittedChains` permissions into
+ * an equivalent CAIP-25 permission.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly
+ * what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by
+ * controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(state: Record<string, unknown>) {
+  if (
+    !hasProperty(state, 'PermissionController') ||
+    !isObject(state.PermissionController) ||
+    !hasProperty(state, 'NetworkController') ||
+    !isObject(state.NetworkController) ||
+    !hasProperty(state, 'SelectedNetworkController') ||
+    !isObject(state.SelectedNetworkController)
+  ) {
+    return state;
+  }
+  const {
+    PermissionController: { subjects },
+    NetworkController: { selectedNetworkClientId, networkConfigurations },
+    SelectedNetworkController: { domains },
+  } = state;
+  if (
+    !isObject(subjects) ||
+    !selectedNetworkClientId ||
+    typeof selectedNetworkClientId !== 'string' ||
+    !isObject(networkConfigurations) ||
+    !isObject(domains)
+  ) {
+    return state;
+  }
+
+  const getChainIdForNetworkClientId = (networkClientId: string) => {
+    const networkConfiguration =
+      (networkConfigurations[networkClientId] as { chainId: Hex }) ??
+      BUILT_IN_NETWORKS[
+        networkClientId as unknown as keyof typeof BUILT_IN_NETWORKS
+      ];
+    return networkConfiguration?.chainId;
+  };
+
+  const currentChainId = getChainIdForNetworkClientId(selectedNetworkClientId);
+  if (!currentChainId || typeof currentChainId !== 'string') {
+    return state;
```

Comment:
there are a bunch of places where we're returning because of malformed state. Wondering if we should captureExceptions in these cases

Reply:
added https://github.com/MetaMask/metamask-extension/pull/26483/commits/35f80921d6f6e90de5051f227b2c5c948088e5b5

Reply:
Sweet! Thanks!